addons/controllers/cpi/oraclecpiconfig_controller.go

```diff
@@ -0,0 +1,240 @@
+// the owner cluster and the authentication secret are required to reconcile the OracleCPIConfig
+func (r *OracleCPIConfigReconciler) reconcileOracleCPIConfig(ctx context.Context, cpiConfig *cpiv1alpha1.OracleCPIConfig,
+	cluster *clusterapiv1beta1.Cluster, auth *v1.Secret) (_ ctrl.Result, retErr error) {
+	// patch the CPIConfig CR in the end
+	patchHelper, err := clusterapipatchutil.NewHelper(cpiConfig, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// patch OracleCPIConfig before returning the function
+	defer func() {
+		r.Log.Info("Patching OracleCPIConfig")
+		if err := patchHelper.Patch(ctx, cpiConfig); err != nil {
+			r.Log.Error(err, "Error patching OracleCPIConfig")
+			retErr = err
+		}
+		r.Log.Info("Successfully patched OracleCPIConfig")
+	}()
+
+	if !cpiConfig.GetDeletionTimestamp().IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	fingerprint, ok := auth.Data["fingerprint"]
+	if !ok {
+		r.Log.Info("Cannot extract fingerprint", "name", authenticationSecretName, "namespace", authenticationSecretNamespace)
+	}
+	key, ok := auth.Data["key"]
+	if !ok {
+		r.Log.Info("Cannot extract key", "name", authenticationSecretName, "namespace", authenticationSecretNamespace)
+	}
+	region, ok := auth.Data["region"]
+	if !ok {
+		r.Log.Info("Cannot extract region", "name", authenticationSecretName, "namespace", authenticationSecretNamespace)
+	}
+	tenancy, ok := auth.Data["tenancy"]
+	if !ok {
+		r.Log.Info("Cannot extract tenancy", "name", authenticationSecretName, "namespace", authenticationSecretNamespace)
+	}
+	user, ok := auth.Data["user"]
+	if !ok {
+		r.Log.Info("Cannot extract user", "name", authenticationSecretName, "namespace", authenticationSecretNamespace)
+	}
+	// the passphrase is optional, use zero value if not provided
+	passphrase := auth.Data["passphrase"]
+	compartment, err := util.ParseClusterVariableString(cluster, "compartmentId")
+	if err != nil {
+		r.Log.Error(err, "Cannot extract compartment from cluster", "cluster", cluster.Name)
+	}
+	vcn, err := util.ParseClusterVariableString(cluster, "externalVCNId")
+	if err != nil {
+		r.Log.Error(err, "Cannot extract vcn from cluster", "cluster", cluster.Name)
+	}
+
+	subnet, err := util.ParseClusterVariableString(cluster, "privateServiceSubnetId")
+	if err != nil {
+		r.Log.Error(err, "Cannot extract private subnet from cluster", "cluster", cluster.Name)
+	}
+
+	// convert the CPIConfig CR to data values
+	d := &OracleCPIDataValues{
+		Auth: OracleCPIDataValuesAuth{
+			Region:      string(region),
+			Tenancy:     string(tenancy),
+			User:        string(user),
+			Key:         string(key),
+			Fingerprint: string(fingerprint),
+			Passphrase:  string(passphrase),
+		},
+		Compartment: compartment,
+		VCN:         vcn,
+		LoadBalancer: struct {
+			Subnet1 string `yaml:"subnet1"`
+			Subnet2 string `yaml:"subnet2"`
+		}{
+			Subnet1: subnet,
+			Subnet2: subnet,
+		},
+	}
+
+	// generate data value secrets for the Oracle CPI package
+	secret := &v1.Secret{
```

Comment:
nit: You can assemble secret object here, which will make the code more readable

Reply:
stringData, ownerReference etc.

Reply:
yea its shall be fine for now, and having more consistent styke with the vspherecpiconfig controller